src/Rule/ChangeOnDateRule.php

```diff
@@ -133,25 +133,29 @@ private function getDate($password): ?DateTimeInterface
     private function getMessage(DateConstraint $constraint, DateTimeInterface $date): string
     {
         $translator = Policy::getTranslator();
-        $minString = $constraint->getMin() === null ? '' : $constraint->getMin()->format('Y-m-d H:i:s');
-        $maxString = $constraint->getMax() === null ? '' : $constraint->getMax()->format('Y-m-d H:i:s');
 
-        if ($constraint->getMax() === null) {
+        $constraintMin = $constraint->getMin();
```

Comment:
Doing this call once and storing the result improves performance, though there's very little overhead.

Reply:
My own tool allows you to use this style of coding, but it's off by default because it cannot detect a bug in the following code (not a bug in PHPStan either):

```php
class C {
    /** @var ?int */
    private $x;
    public function getX() : ?int {
        return $this->x;
    }
    public function setX(?int $x) : void {
        $this->x = $x;
    }
}

function takesInt(int $i) : void {}

function takesC(C $c) : void {
    if ($c->getX() !== null) {
        $c->setX(null);
        takesInt($c->getX());
    }
}
```

The code can be made sound by changing the second function to

```php
function takesC(C $c) : void {
    if (($x = $c->getX()) !== null) {
        $c->setX(null);
        takesInt($x);
    }
}
```

though this fails in PHPStan because [assignments in conditionals do not work](https://github.com/phpstan/phpstan/issues/701)

Reply:
Ok, so you suggested change is not really a speed optimization, but rather to make static analysis happy since proving that `$constraint->getMin()` and `$constraint->getMax()` return the same type every time they are called cannot be done?

Reply:
Yes, agreed. But let's consider the issue as reported by static analysis for a moment.

The following code is unsafe, since there is no way to guarantee that `$constraint->getMin()` doesn't return `null` inside the if-statement.

```php
if ($constraint->getMin() !== null) {
    $constraint->getMin()->format(DateTime::RFC3339_EXTENDED);
}
```

The following code gets around the issue (as suggested by this PR).

```php
$constraintMin = $constraint->getMin();
if ($constraintMin !== null) {
    $constraintMin->format(DateTime::RFC3339_EXTENDED);
}
```

From a static analysis point-of-view, the second code example is less error-prone. But if there was an error in the first code example, it would be due to a bug in `$constraint->getMin()`, and the likelihood of such a bug is still the same in the second code example - it is just a bit less likely to be discovered.

With a bug so that `$constraint->getMin()` first returns a `DateTime` and then `null` on consecutive calls, PHP would produce a fatal error in the first code example, and the bug would be exposed. In the second code example, however, the bug would not be discovered.

To me, the first code example seems best for development and testing, and then the second code example would be safest in production, where we don't want a fatal error.

Do you have any thoughts on the matter? Is this just overthinking it? :)

Reply:
It is not the responsibility of code using `$constraint->getMin()` to expose bugs in the method. So I guess I was just overthinking.